Fixed calculation of expectation in SingleContinousDistribution when expression is zero

kc611 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #20756

#### Brief description of what is fixed or changed

The issue was arising due to `SingleContinuousDistribution` trying to calculate expectation using Taylor series. 

Since degree of `Poly(0)` is `-oo` It tries to calculate series of values the mgf over infinite points.

https://github.com/sympy/sympy/blob/ab39cbb48e23284889ca851c35845ec9b66a37c9/sympy/stats/crv.py#L427-L428

Rather, It'll be better for it to follow the simple integration path when expression is zero. The case in which it'll simply return zero for integrating over zero.   

https://github.com/sympy/sympy/blob/ab39cbb48e23284889ca851c35845ec9b66a37c9/sympy/stats/crv.py#L436

#### Other comments


#### Release Notes


<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

- stats
  - Expectation can now be calculated across multiple Uniform RVs evaluating to zero

<!-- END RELEASE NOTES -->

oscarbenjamin commented on sympy/stats/crv.py:
```diff
@@ -417,7 +417,7 @@ def moment_generating_function(self, t, **kwargs):
     def expectation(self, expr, var, evaluate=True, **kwargs):
         """ Expectation of expression over distribution """
-        if evaluate:
+        if evaluate and expr:
```
Testing `if expr` like this will not catch all cases e.g.:
```python
In [4]: e = x*(x + 1) - x*x - x

In [5]: e
Out[5]: 
   2                
- x  + x⋅(x + 1) - x

In [6]: bool(e)
Out[6]: True

In [7]: expand(e)
Out[7]: 0

In [8]: Poly(e, x)
Out[8]: Poly(0, x, domain='ZZ')

In [9]: Poly(e, x).degree()
Out[9]: -∞
```
If would be better to test if `p` is zero.

A test for a case like this should be added.

oscarbenjamin commented on sympy/stats/crv.py:
```diff
@@ -417,7 +417,7 @@ def moment_generating_function(self, t, **kwargs):
     def expectation(self, expr, var, evaluate=True, **kwargs):
         """ Expectation of expression over distribution """
-        if evaluate:
+        if evaluate and expr.is_zero is False:
```
This will also fail in some cases. What if `is_zero` gives `None`?

You should check `p` below instead (as I suggested before).

kc611 replied:
> You should check p below instead

Wouldn't `p` suffer with the same problem ? Or is it that `is_zero` doesn't give `None` for `poly` expressions.  Besides isn't the `Integral` doing same thing (mathematically) as the `try - except` block ?

oscarbenjamin replied:
Poly are very different from Expr:
```
In [14]: e = x*(x + 1) - x*x - x

In [15]: e
Out[15]: 
   2                
- x  + x⋅(x + 1) - x

In [16]: Poly(e, x)
Out[16]: Poly(0, x, domain='ZZ')
```
The `else` block returns an unevaluated `Integral` which is not what is wanted when `evaluate=True` (compare `integrate` vs `Integral`).

kc611 replied:
> compare integrate vs Integral 

Ah, I see. 

So you meant adding something like : 

```
if p.is_zero:
__return integrate(expr * self.pdf(var), (var, self.set), **kwargs)
```
After p has been evaluated. Will that do?

oscarbenjamin replied:
Each time we discuss a particular case like this you should consider whether or not there is a test that covers that case. If there isn't then one should be added.

The change I am commenting on here would have caused an unevaluated integral to be returned whenever `expr.is_zero` gives `None`. Is there a test for that case?

kc611 replied:
Not that I know of. Expectation calculation for `SingleContinuousDistribution` takes in expressions of type distributions which `is_zero` check almost always gives `False` Since the distributions almost always have a value (or rather a variable that contains the distribution which is always regarded as non-zero ). So regardless of how small we make their coefficients or how tangled we make make that equation it gives `is_zero` check as `False`. The other possibility was the expression being exactly zero (as is in the issue). 

If you encounter any such case for which an expression having `is_zero` as `None` is being passed here please mention one here so it could be handled appropriately.

oscarbenjamin replied:
I gave an example:
```python
In [1]: e = x*(x + 1) - x*x - x

In [2]: e
Out[2]: 
   2                
- x  + x⋅(x + 1) - x

In [3]: print(e.is_zero)
None
```

kc611 replied:
But currently we have added the `is_zero` check after the conversion into `Poly` for which this particular case also evaluates to `True`. Moreover I think the expression is also being simplified before entering this function. So direct implementation of any such example isn't working.

oscarbenjamin replied:
Part of the reason for adding tests is to test the current behaviour of the new code with the changes made in any particular PR. The more important reason though is to test the future behaviour when someone else *changes* the code later. If a future contributor tried to make the change that you did and test `expr.is_zero` then the test failure would show them why that is wrong. If the code that you showed before was able to pass the test suite then that means that there are not enough tests to pick up on this potential problem.

That is why each time you discover a problem with the code in a PR you should not only fix that problem in the code but also *add a test* (unless there already is one). The test makes sure that the code works correctly *now*, and it proves to anyone reviewing the PR that it correctly handles the tested examples. More importantly it ensures that those examples will be handled correctly *in future* even if the code is changed.

kc611 replied:
Thanks, I'll keep this in mind.

oscarbenjamin commented on sympy/stats/crv.py:
```diff
@@ -422,7 +422,7 @@ def expectation(self, expr, var, evaluate=True, **kwargs):
                 mgf = self._moment_generating_function(t)
-                if mgf is None:
+                if mgf is None or p.is_zero:
                     return integrate(expr * self.pdf(var), (var, self.set), **kwargs)
```
If `p` is zero then that implies that `expr` is zero so is there any need to call `integrate` in that case?

Can we not just return zero?

kc611 replied:
That'll be better. I'll make the necessary change.

oscarbenjamin committed:
fix(stats): fix expectation of zero expression

Previously the expectation operator E(X*Y) would fail in some cases when
the expression had an expectation of zero.

Co-authored-by: kc611 <ckaustubhm06@gmail.com>

oscarbenjamin commented:
I've squashed the commits.

Looks good to merge once tests are passed.
